File: BOJ/seulah/6_week/컨베이어 벨트 위의 로봇.py
import sys
from collections import deque
input = sys.stdin.readline

N,K = map(int,input().split())
belt = deque(map(int,input().split()))
robots = deque([0]*(2*N))
step = 0

while True:
    step += 1

    # 1) 벨트 및 로봇 회전
    belt.rotate(1)
    robots.rotate(1)
    robots[N - 1] = 0 # 로봇 내리기

    # 2) 로봇 이동
    for i in range(N - 2, -1, -1): # 가장 먼저 올린 로봇부터 이동하기 위해 역순 탐색
        if(robots[i] and not robots[i+1] and belt[i+1]>=1):
            robots[i+1] = 1
            robots[i] = 0

            belt[i+1] -=1 # 내구도 감소
    robots[N - 1] = 0 # 로봇 내리기

    # 3) 로봇 올리기
    if belt[0] >=1 and not robots[0]:
        robots[0] =1
        belt[0] -= 1 # 내구도 감소

    # 4) 내구도 0 개수 세기
    if belt.count(0) >= K:
        break

print(step)

File: BOJ/jaehong/6_week/컨베이어 벨트 위의 로봇.js
const input = require('fs')
  .readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/example.txt')
  .toString().trim().split('\n').map(e => e.split(' ').map(Number));

const [n, k] = input.shift();
const belt = input.shift();
const robots = Array.from({ length: 2 * n }, () => false);
let stage = 0;

const check = () => {
  const val = belt.reduce((acc, cur) => acc + (cur === 0 ? 1 : 0), 0);
  return val
}

const rotate = () => {
  const temp = belt.pop();
  belt.unshift(temp);

  robots.pop();
  robots.unshift(false);
}

const remove = () => {
  robots[n - 1] = false;
}

const moveRo = () => {
  for (let i = n - 2; i > -1; i--){
    if (!robots[i]) continue;

    if (!robots[i + 1] && belt[i + 1] > 0) {
      robots[i] = false;
      robots[i + 1] = true;
      belt[i + 1] -= 1;
    }
  }
}

const add = () => {
  if (belt[0] > 0) {
    belt[0] -= 1;
    robots[0] = true;
  }
}

const solution = () => {

  while (check() < k) {
    stage++;

    rotate();
    if (robots[n - 1]) remove();

    moveRo();
    if (robots[n - 1]) remove();
    add();
  }
  console.log(stage)
}
solution();